src/server/terrain/simple_structures/SbbGen.zig

```diff
@@ -0,0 +1,93 @@
+const std = @import("std");
+
+const main = @import("main");
+const terrain = main.server.terrain;
+const Vec3i = main.vec.Vec3i;
+const GenerationMode = terrain.biomes.SimpleStructureModel.GenerationMode;
+const CaveMapView = terrain.CaveMap.CaveMapView;
+const CaveBiomeMapView = terrain.CaveBiomeMap.CaveBiomeMapView;
+const sbb = terrain.structure_building_blocks;
+const Blueprint = main.blueprint.Blueprint;
+const ZonElement = main.ZonElement;
+const Neighbor = main.chunk.Neighbor;
+const ServerChunk = main.chunk.ServerChunk;
+const NeverFailingAllocator = main.heap.NeverFailingAllocator;
+const parseBlock = main.blocks.parseBlock;
+
+pub const id = "cubyz:sbb";
+pub const generationMode = .floor;
+
+const SbbGen = @This();
+
+structure: []const u8,
+structureRef: *const sbb.StructureBuildingBlock,
+placeMode: Blueprint.PasteMode,
+
+pub fn getHash(self: SbbGen) u64 {
+	return std.hash.Wyhash.hash(@intFromEnum(self.placeMode), self.structure);
+}
+
+pub fn loadModel(arenaAllocator: NeverFailingAllocator, parameters: ZonElement) *SbbGen {
+	const structureId = parameters.get(?[]const u8, "structure", null) orelse unreachable;
+	const structureRef = sbb.getByStringId(structureId) orelse {
+		std.log.err("Could not find structure building block with id '{s}'", .{structureId});
+		unreachable;
+	};
+	const self = arenaAllocator.create(SbbGen);
+	self.* = .{
+		.structure = structureId,
+		.structureRef = structureRef,
+		.placeMode = std.meta.stringToEnum(Blueprint.PasteMode, parameters.get([]const u8, "placeMode", "degradable")) orelse Blueprint.PasteMode.degradable,
+	};
+	return self;
+}
+
+pub fn generate(self: *SbbGen, _: GenerationMode, x: i32, y: i32, z: i32, chunk: *ServerChunk, _: CaveMapView, _: CaveBiomeMapView, seed: *u64, _: bool) void {
```

Comment:
For the future: It might be worth exploring to use the CaveMap here.
E.g. we could use it to only generate buildings in a reasonably flat area, or check if there is enough open space for a tree to grow..

Reply:
#1357

Not that trivial, you first have to resolve all children before you know how high the structure will be.

Reply:
Or, you just let the addon creator give you a bounding box. It doesn't really need to be precise anyways.

Reply:
I mean, you can calculate upper bound during loading (when resolving references to be precise) without making structure creators update height of their structures every time they modify one segment of the structure. Just pick tallest blueprint for each child block. Technically you could also just resolve the whole structure while placing it and discard it if it is too tall, but that is quite some waste.

Reply:
But what if the addon creator wants to allow some leaves cut off by the cave ceiling, but leave enough room for the stem?

Reply:
Alr, I guess they will have to do maths then